src/ORM/DataObjectSchema.php

```diff
@@ -486,29 +500,65 @@ protected function cacheDatabaseFields($class)
     }
 
     /**
-     * Cache all indexes for the given class.
-     * Will do nothing if already cached
+     * Cache all indexes for the given class. Will do nothing if already cached.
      *
      * @param $class
      */
     protected function cacheDatabaseIndexes($class)
     {
-        if (array_key_exists($class, $this->databaseIndexes)) {
-            return;
+        if (!array_key_exists($class, $this->databaseIndexes)) {
+            $this->databaseIndexes[$class] = array_merge(
+                $this->defaultDatabaseIndexes($class),
+                $this->customDatabaseIndexes($class)
+            );
         }
+    }
+
+    /**
+     * Get "default" database indexable field types
+     *
+     * @param  string $class
+     * @return array
+     */
+    protected function defaultDatabaseIndexes($class)
+    {
         $indexes = [];
+        if (empty($this->defaultDatabaseIndexes[$class])) {
+            $this->defaultDatabaseIndexes[$class] = [];
+        }
 
-        // look for indexable field types
-        foreach ($this->databaseFields($class, false) as $field => $type) {
-            if ($type === 'ForeignKey' || $type === 'DBClassName') {
-                $indexes[$field] = [
-                    'type' => 'index',
-                    'columns' => [$field],
-                ];
+        // Ensure that fixed field types, e.g. ClassName can be indexed
+        $databaseFields = $this->databaseFields($class, false);
+        foreach ($databaseFields as $name => $type) {
+            /** @var DBField $fieldObj */
+            $fieldObj = Injector::inst()->create($type, $name);
+            if ($fieldObj->config()->get('indexed')) {
+                $this->defaultDatabaseIndexes[$class][] = $name;
             }
         }
 
-        // look for custom indexes declared on the class
+        // Return the registered default indexes
+        foreach ($this->defaultDatabaseIndexes[$class] as $field) {
+            $indexes[$field] = [
+                'type' => 'index',
+                'columns' => [$field],
+            ];
+        }
+
+        return $indexes;
+    }
+
+    /**
+     * Look for custom indexes declared on the class
+     *
+     * @param  string $class
+     * @return array
+     * @throws InvalidArgumentException If an index already exists on the class
+     * @throws InvalidArgumentException If a custom index format is not valid
+     */
+    protected function customDatabaseIndexes($class)
```

Comment:
As with `buildTableName` being called from `cacheTableNames`, this should be called `buildCustomDatabaseIndexes` as it's called from `cacheDatabaseIndexes`, so that it doesn't look like the standard getter.